File: content/problems/1801-number-of-orders-in-the-backlog/solutions/dual-price-heaps/Solution.kt
class Solution {
    fun getNumberOfBacklogOrders(orders:Array<IntArray>):Int{
        val buy=java.util.PriorityQueue<IntArray>{
            a,b->b[0]-a[0]
        };
        val sell=java.util.PriorityQueue<IntArray>(compareBy{
            it[0]
        });
        for(order in orders){
            val price=order[0];
            var amount=order[1];
            if(order[2]==0){
                while(amount>0&&sell.isNotEmpty()&&sell.peek()[0]<=price){
                    val other=sell.peek();
                    val traded=minOf(amount,other[1]);
                    amount-=traded;
                    other[1]-=traded;
                    if(other[1]==0)sell.remove()
                };
                if(amount>0)buy.add(intArrayOf(price,amount))
            }else{
                while(amount>0&&buy.isNotEmpty()&&buy.peek()[0]>=price){
                    val other=buy.peek();
                    val traded=minOf(amount,other[1]);
                    amount-=traded;
                    other[1]-=traded;
                    if(other[1]==0)buy.remove()
                };
                if(amount>0)sell.add(intArrayOf(price,amount))
            }
        };
        return ((buy.sumOf{
            it[1].toLong()
        }+sell.sumOf{
            it[1].toLong()
        })%1_000_000_007).toInt()
    }
}

File: content/problems/1801-number-of-orders-in-the-backlog/solutions/dual-price-heaps/solution.cpp
class Solution {
    public: int getNumberOfBacklogOrders(vector<vector<int>>& orders){
        priority_queue<pair<int,int>> buy;
        priority_queue<pair<int,int>,vector<pair<int,int>>,greater<pair<int,int>>> sell;
        for(auto order:orders){
            int price=order[0],amount=order[1];
            if(!order[2]){
                while(amount&&!sell.empty()&&sell.top().first<=price){
                    auto [otherPrice,otherAmount]=sell.top();
                    sell.pop();
                    int traded=min(amount,otherAmount);
                    amount-=traded;
                    otherAmount-=traded;
                    if(otherAmount)sell.push({
                        otherPrice,otherAmount
                    });
                }
                if(amount)buy.push({
                    price,amount
                });
            }else{
                while(amount&&!buy.empty()&&buy.top().first>=price){
                    auto [otherPrice,otherAmount]=buy.top();
                    buy.pop();
                    int traded=min(amount,otherAmount);
                    amount-=traded;
                    otherAmount-=traded;
                    if(otherAmount)buy.push({
                        otherPrice,otherAmount
                    });
                }
                if(amount)sell.push({
                    price,amount
                });
            }
        }
        long long total=0;
        while(!buy.empty()){
            total+=buy.top().second;
            buy.pop();
        }while(!sell.empty()){
            total+=sell.top().second;
            sell.pop();
        }
        return total%1000000007;
    }
};
